<a href="https://colab.research.google.com/github/mudesir/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Mudesir_of_DS16_Unit_1_Sprint_Challenge_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 2

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 is a variable that indicates whether or not a person's request for credit has been approved or denied. This is a good candidate for a y variable since we might want to use the other features to predict this one. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [1]:
# load the data
import pandas as pd
import numpy as np
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data')
print(df.shape)
df.head()

(689, 16)


b  30.83      0  u  g  w  v  1.25  t t.1  01  f g.1  00202  0.1  +
0  a  58.67  4.460  u  g  q  h  3.04  t   t   6  f   g  00043  560  +
1  a  24.50  0.500  u  g  q  h  1.50  t   f   0  f   g  00280  824  +
2  b  27.83  1.540  u  g  w  v  3.75  t   t   5  t   g  00100    3  +
3  b  20.17  5.625  u  g  w  v  1.71  t   f   0  f   s  00120    0  +
4  b  32.08  4.000  u  g  m  v  2.50  t   f   0  t   g  00360    0  +

In [2]:
column_headers = ['A1', 'A2','A3','A4','A5', 'A6','A7','A8','A9', 'A10','A11','A12','A13', 'A14','A15','A16']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data', names=column_headers)
print(df.shape)
df.head()

(690, 16)


A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

In [0]:
df.isnull().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

In [3]:
df = df.replace({'?': np.NaN})
print(df.shape)
df.head()

(690, 16)


A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

In [4]:
df.isnull().sum()


A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

In [5]:
df.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

In [18]:
df['A2'].value_counts(dropna=False, ascending=False)[:20]

NaN      12
22.67     9
20.42     7
22.50     6
23.58     6
24.50     6
19.17     6
20.67     6
25.00     6
18.83     6
23.08     5
33.17     5
23.00     5
27.83     5
27.67     5
23.25     5
25.17     4
28.58     4
24.58     4
22.92     4
Name: A2, dtype: int64

In [19]:
df['A14'].value_counts(dropna=False, ascending=False)[:20]

00000    132
00200     35
00120     35
00160     34
00080     30
00100     30
00280     22
00180     18
00140     16
00320     14
00240     14
NaN       13
00300     13
00260     11
00220      9
00060      9
00400      9
00340      7
00360      7
00380      5
Name: A14, dtype: int64

In [50]:
df.dropna(inplace=True)
df.isnull().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

In [0]:
# we have to find for some continus value it become object type 'A2' And 'A14'
def int_rate_to_float(cell_contents):
  return float(cell_contents)

In [52]:
# Apply our solution to every cell in a column
df['A2'] = df['A2'].apply(int_rate_to_float)
df.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +

In [53]:
df['A14'] = df['A14'].apply(int_rate_to_float)
df.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +

In [54]:
df.describe()

A2          A3  ...          A14            A15
count  653.000000  653.000000  ...   653.000000     653.000000
mean    31.503813    4.829533  ...   180.359877    1013.761103
std     11.838267    5.027077  ...   168.296811    5253.278504
min     13.750000    0.000000  ...     0.000000       0.000000
25%     22.580000    1.040000  ...    73.000000       0.000000
50%     28.420000    2.835000  ...   160.000000       5.000000
75%     38.250000    7.500000  ...   272.000000     400.000000
max     76.750000   28.000000  ...  2000.000000  100000.000000

[8 rows x 6 columns]

In [55]:
df.describe(exclude='number')

A1   A4   A5   A6   A7   A9  A10  A12  A13  A16
count   653  653  653  653  653  653  653  653  653  653
unique    2    3    3   14    9    2    2    2    3    2
top       b    u    g    c    v    t    f    f    g    -
freq    450  499  499  133  381  349  366  351  598  357

In [56]:
df.describe(include="all")

A1          A2          A3   A4  ...  A13          A14            A15  A16
count   653  653.000000  653.000000  653  ...  653   653.000000     653.000000  653
unique    2         NaN         NaN    3  ...    3          NaN            NaN    2
top       b         NaN         NaN    u  ...    g          NaN            NaN    -
freq    450         NaN         NaN  499  ...  598          NaN            NaN  357
mean    NaN   31.503813    4.829533  NaN  ...  NaN   180.359877    1013.761103  NaN
std     NaN   11.838267    5.027077  NaN  ...  NaN   168.296811    5253.278504  NaN
min     NaN   13.750000    0.000000  NaN  ...  NaN     0.000000       0.000000  NaN
25%     NaN   22.580000    1.040000  NaN  ...  NaN    73.000000       0.000000  NaN
50%     NaN   28.420000    2.835000  NaN  ...  NaN   160.000000       5.000000  NaN
75%     NaN   38.250000    7.500000  NaN  ...  NaN   272.000000     400.000000  NaN
max     NaN   76.750000   28.000000  NaN  ...  NaN  2000.000000  100000.000000  NaN

[11 rows x 16 columns]

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

1. Null Hypothesis the two means('+' and '-') are equal
2. Alternative the two means are not equal
3. Confidence Interval: 95%


In [0]:
b_acept = df[df['A16']=='+']
b_rejec = df[df['A16']=='-']

In [61]:
#Test A8 and A16
from scipy import stats
stats.ttest_ind(b_rejec['A8'], b_acept['A8'], nan_policy='omit')

Ttest_indResult(statistic=-9.002392498622463, pvalue=2.4079238505859142e-18)

Conclusion: Given t statistic -9.0 and p-value 0, we reject the null hypothesis that, the two means are equal, we suggest the alternative hypothesis that the two means are not equal.

In [63]:
#Test A2 and A16
stats.ttest_ind(b_rejec['A2'], b_acept['A2'])

Ttest_indResult(statistic=-4.675662433767456, pvalue=3.5636710933835225e-06)

Conclusion: Given t statistic -4.68 and p-value 0, we reject the null hypothesis that, the two means are equal, we suggest the alternative hypothesis that the two means are not equal.

In [65]:
#Test A14 and A16
stats.ttest_ind(b_rejec['A14'], b_acept['A14'])

Ttest_indResult(statistic=2.18221936438708, pvalue=0.029450100409286473)

Conclusion: Given t statistic 2.18 and p-value 0.0295, we fail to reject the null hypothesis that, the two means are equal.

To chi2 test
1. Null hypothesis the two variables are independent
2. Alternative the two variables are not independent
3. Confidence level: 95%

In [67]:

df = df.replace({'t': 1, 'f': 0})
print(df.shape)
df.head()


(653, 16)


A1     A2     A3 A4 A5 A6 A7    A8  A9  A10  A11  A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25   1    1    1    0   g  202.0    0   +
1  a  58.67  4.460  u  g  q  h  3.04   1    1    6    0   g   43.0  560   +
2  a  24.50  0.500  u  g  q  h  1.50   1    0    0    0   g  280.0  824   +
3  b  27.83  1.540  u  g  w  v  3.75   1    1    5    1   g  100.0    3   +
4  b  20.17  5.625  u  g  w  v  1.71   1    0    0    0   s  120.0    0   +

In [68]:
df = df.replace({'b': 1, 'a': 0})
print(df.shape)
df.head()

(653, 16)


A1     A2     A3 A4 A5 A6 A7    A8  A9  A10  A11  A12 A13    A14  A15 A16
0   1  30.83  0.000  u  g  w  v  1.25   1    1    1    0   g  202.0    0   +
1   0  58.67  4.460  u  g  q  h  3.04   1    1    6    0   g   43.0  560   +
2   0  24.50  0.500  u  g  q  h  1.50   1    0    0    0   g  280.0  824   +
3   1  27.83  1.540  u  g  w  v  3.75   1    1    5    1   g  100.0    3   +
4   1  20.17  5.625  u  g  w  v  1.71   1    0    0    0   s  120.0    0   +

In [80]:
cidata2 = pd.crosstab(df['A16'],df['A1'])
cidata2

A1     0    1
A16          
+     95  201
-    108  249

In [83]:
chi2, P_value, dof, expected = stats.chi2_contingency(cidata2)
print('chi2 ', chi2)
print('P-value ', P_value)
print('dof ', dof)
print('expected ', expected)


chi2  0.17764082160253514
P-value  0.6734085695133722
dof  1
expected  [[ 92.01837672 203.98162328]
 [110.98162328 246.01837672]]


Conclusion:Due to a chi2 value of 0.1776 and p-value 0.67 we fail to reject that the two variables are independent.

In [78]:
cidata3 = pd.crosstab(df['A16'],df['A4'])
cidata3

A4   l    u    y
A16             
+    2  249   45
-    0  250  107

In [79]:
chi2, P_value, dof, expected = stats.chi2_contingency(cidata3)
print('chi2 ', chi2)
print('P-value ', P_value)
print('dof ', dof)
print('expected ', expected)

chi2  21.78325079317282
P-value  1.8613463470618034e-05
dof  2
expected  [[  0.90658499 226.19295559  68.90045942]
 [  1.09341501 272.80704441  83.09954058]]


Conclusion:Due to a chi2 value of 21.783 and p-value 0.000015 we reject that the two variables are independent, and suggest the alternative they are related or dependent each other.

In [75]:
cidata4 = pd.crosstab(df['A16'], df['A6'])
cidata4

A6   aa   c  cc   d   e  ff   i  j   k   m   q  r   w   x
A16                                                      
+    19  60  29   7  14   7  14  3  13  16  49  2  33  30
-    33  73  11  19  10  43  41  7  35  22  26  1  30   6

In [76]:
chi2, P_value, dof, expected = stats.chi2_contingency(cidata4)
print('chi2 ', chi2)
print('P-value ', P_value)
print('dof ', dof)
print('expected ', expected)

chi2  89.76481160702343
P-value  1.5500154549498966e-13
dof  13
expected  [[23.5712098  60.28790199 18.13169985 11.7856049  10.87901991 22.66462481
  24.93108729  4.53292496 21.75803982 17.22511485 33.99693721  1.35987749
  28.55742726 16.31852986]
 [28.4287902  72.71209801 21.86830015 14.2143951  13.12098009 27.33537519
  30.06891271  5.46707504 26.24196018 20.77488515 41.00306279  1.64012251
  34.44257274 19.68147014]]


Conclusion:Due to a chi2 value of 89.76 and p-value 0 we reject that the two variables are independent, and suggest the alternative they are related or dependent each other.

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.



*   Infact I did more than 2 tests, the t-tests tells me that for some categories you reject the null hypothesis that the mean of the 2 compard samples are not the same in other way whene i compare A14 and A16 I fail to reject the null hypothesis that the two sample have equal mean. Inother word what ever feature and/or class owned by the variable we will inference them using statistical tools.
*   I take 3 chi2 tests I gave a conclussion for each of them. as general the chi2 test tells us the dependent and independent nature of the given variables from my 3 test I fail to reject one of the categorical feature and I reject 2 of them.


*   The most challenging of this sprint challenge is the given data not readable or understandabl we can not tell anything about the features and class, just telling number so this affect your conclussion and analysis.

*   Time for study





## Part 4 - Bayesian vs Frequentist Statistics

Using a minimum of 2-3 sentences, give an explanation of Bayesian and Frequentist statistics, and then compare and contrast these two approaches to statistical inference.



The Bayesian is subjective and uses a priori beliefs to define a prior probability distribution on the possible values of the unknown parameters. ... The frequentist see probability as something that has to do with a limiting frequency based on an observed proportion

# Stretch Goals: 
Do these to get a 3. These are not required in order to pass the Sprint Challenge.

## Part 1: 

Make sure that all of your dataframe columns have the appropriate data types. *Hint:* If a column has the datatype of "object" even though it's made up of float or integer values, you can coerce it to act as a numeric column by using the `pd.to_numeric()` function. In order to get a 3 on this section make sure that your data exploration is particularly well commented, easy to follow, and thorough.

## Part 2:

Write functions that can calculate t-tests and chi^2 tests on all of the appropriate column combinations from the dataset. (Remember that certain tests require certain variable types.)

## Part 3: 

Calculate and report confidence intervals on your most important mean estimates (choose at least two). Make some kind of a graphic or visualization to help us see visually how precise these estimates are.

## Part 4:

Give an extra awesome explanation of Bayesian vs Frequentist Statistics. Maybe use code or visualizations, or any other means necessary to show an above average grasp of these high level concepts.

In [0]:
# You can work the stretch goals down here or back up in their regular sections
# just make sure that they are labeled so that we can easily differentiate
# your main work from the stretch goals.